aiida/tools/graph/graph_traversers.py

```diff
@@ -0,0 +1,287 @@
+
+    valid_output = {
+        'rules_applied': rules_applied,
+        'forward': links_forward,
+        'backward': links_backward,
+    }
+    return valid_output
+
+
+def traverse_graph(starting_pks, max_iterations=None, get_links=False, links_forward=(), links_backward=()):
+    """
+    This function will return the set of all nodes that can be connected
+    to a list of initial nodes through any sequence of specified links.
+    Optionally, it may also return the links that connect these nodes.
+
+    :type starting_pks: list or tuple
+    :param starting_pks: The set of (valid) pks for the starting nodes (also accepts sets).
+
+    :type max_iterations: int or None
+    :param max_iterations:
+        The number of iterations to apply the set of rules (a value of 'None' will
+        iterate until no new nodes are added).
+
+    :type get_links:  bool
+    :param get_links:
+        Pass True to also return the links between all nodes (found + initial).
+
+    :type links_forward: aiida.common.links.LinkType
+    :param links_forward:
+        List with all the links that should be traversed in the forward direction.
+
+    :type links_backward: aiida.common.links.LinkType
+    :param links_backward:
+        List with all the links that should be traversed in the backward direction.
+    """
+    # pylint: disable=too-many-locals,too-many-statements,too-many-branches
+    from aiida import orm
+    from aiida.tools.graph.age_entities import Basket
+    from aiida.tools.graph.age_rules import UpdateRule, RuleSequence, RuleSaveWalkers, RuleSetWalkers
+    from aiida.common import exceptions
+
+    if max_iterations is None:
+        max_iterations = inf
+    elif not (isinstance(max_iterations, int) or max_iterations is inf):
+        raise TypeError('Max_iterations has to be an integer or infinity')
+
+    linktype_list = []
+    for linktype in links_forward:
+        if not isinstance(linktype, LinkType):
+            raise TypeError('links_forward should contain links, but one of them is: {}'.format(type(linktype)))
+        linktype_list.append(linktype.value)
+    filters_forwards = {'type': {'in': linktype_list}}
+
+    linktype_list = []
+    for linktype in links_backward:
+        if not isinstance(linktype, LinkType):
+            raise TypeError('links_backward should contain links, but one of them is: {}'.format(type(linktype)))
+        linktype_list.append(linktype.value)
+    filters_backwards = {'type': {'in': linktype_list}}
+
+    if not isinstance(starting_pks, (list, set, tuple)):
+        raise TypeError('starting_pks must be of type list, set or tuple\ninstead, it is {}'.format(type(starting_pks)))
+
+    if not starting_pks:
+        if get_links:
+            output = {'nodes': set(), 'links': set()}
+        else:
+            output = {'nodes': set(), 'links': None}
+        return output
+
+    operational_set = set()
+    for single_pk in starting_pks:
+        if not isinstance(single_pk, int):
```

Comment:
How about the following
```
if any([not isinstance(pk, int) for pk in starting_pks]):
    raise
```
Probably this is faster most of the time, because you can also just do `set(starting_pks)` instead of adding them one by one.